transcom/mymove
# B-20767 EDI Error Display Int

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20767)

## Summary

We communicate with TPPS (Third Party Payment System) with a series of reports. When we send/receive/process these reports, we have some error logging that we do as well as some updates to the payment request status throughout the process.

Reports: 

- EDI 858: MilMove generates this report and sends to TPPS
   - if the send fails, PR status = `EDI_ERROR`
   - If the send fails, we also log this error in the `edi_errors` table
   - if the send is successful, PR status = `SENT_TO_GEX` 
- EDI 824
   - We receive this report from TPPS for every payment request, it contains information about errors with the 858.
   - We process this report, and if any errors are found, we change the status of the PR to `EDI_ERROR`
   - If we find errors, we also log this error in the `edi_errors` table
- EDI 997
  - We receive this report from TPPS, it confirms that TPPS had a successful receipt of the 858.
  - If we receive this from TPPS, we update the PR status to `TPPS_RECEIVED`
  - We process this report, and if any errors are found, we do NOT change the status of the PR to `EDI_ERROR`
   - If we find errors with the 997, we also log this error in the `edi_errors` table
- Paid invoice report
  - We will receive this report from TPPS and it will contain information on if/when/amount that TPPS paid HomeSafe.
  - When we receive this report, we will update the PR status to `TPPS_PAID`

In summary, the `edi_errors` table contains error logs for processing 858s, 824s, and 997s. The bulk of the work for this story was to display the edi_errors for a payment request, _regardless of the PR status_, if a value exists in the `edi_errors` table. We want to show the error regardless of status, because for a 997, the PR can have values in the `edi_errors` table, but doesn't necessarily have to be an EDI_ERROR status itself.

If one payment request has multiple associated entries in the edi_errors table, the UI will display the one with the most recent `created_at` date.

Lots of coordination with Dan here to get the UI right but the big elements:
- Error details should only appear IF there is a value in the edi_errors table. If not, nothing mentioning the error details will appear.
- The error details should be collapsible, and not shown by default.
- When there are errors to display, we still want to show the amount that was requested, accepted/rejected, was sent to gex, or was received by gex. I refactored `PaymentRequestCard.jsx` to allow for this.

### How to test

1. Run this script to give yourself some test data (you WILL HAVE TO CHANGE the second value you're inserting for each row, to a valid payment request ID for your local db), or disregard and use your own but be warned that my PR writeup is off of this data.
```
insert into edi_errors
-- 1
VALUES ('b064d62a-9576-402a-b806-41933339c746', 
'154c9ebb-972f-4711-acb2-5911f52aced4', 
null,
'testCode',
'function ProcessReviewedPaymentRequest failed call to edi858c.EDIString: EDI failed validation: Key: ''Invoice858C.Header.ServiceMemberName.ReferenceIdentification'' Error:Field validation for ''ReferenceIdentification'' failed on the ''max'' tag',
'858',
'2024-02-20 21:00:32.37536',
'2024-02-20 21:00:32.37536'),
-- 2
('b064d62a-9576-402a-b806-41933339c747', 
'24c2caf9-9df1-4017-8164-602b24e4c19d', 
null,
'testCode',
'test 824 description',
'824',
'2024-02-22 21:00:32.37536',
'2024-02-22 21:00:32.37536'),
-- 3
('b064d62a-9576-402a-b806-41933339c748', 
'cfd110d4-1f62-401c-a92c-39987a0b4228', 
null,
'testCode',
'test 997 description',
'997',
'2024-02-20 21:00:32.37536',
'2024-02-20 21:00:32.37536')
;
```

3. Login as a TIO.
4. Find a payment request that you did `not` link up to a value in the edi_errors table.
5. Review the payment service items, and get the payment request itself into a "Reviewed" state.
6. You should see this for now: 
![image](https://github.com/user-attachments/assets/b2e99701-24f2-43bf-85f1-bf7be1dc201a)

7. At this point we will simulate the 858 send failure.
8. For the payment request ID that you just reviewed, enter that as the payment request ID for one of the values in the edi_errors table. Also update the payment request status from Reviewed to EDI error.
9. Refresh the page. You will now see the collapsible error details and a status tag of EDI ERROR on the payment request.
10. 
![image](https://github.com/user-attachments/assets/26124ff8-e782-4986-8eda-d81e34002a59)
11. Expand the details and check out the lovely 858 error information.
12. 
![image](https://github.com/user-attachments/assets/9271e2c1-b649-4c27-97e4-fef8598274f5)

13. Now we will simulate that the 858 send worked. Update the status to SENT_TO_GEX and enter a db value for `sent_to_gex_at` and refresh your page. Verify you still see the edi errors details drop down, but that the tag is SENT TO GEX by the PR and you see the date it was sent to gex.
![image](https://github.com/user-attachments/assets/2f6cfc2a-ab36-49e3-9f05-6882ba883d62)
14. The next step would be to simulate TPPS receiving the 858. Change the status to TPPS_RECEIVED and and enter a db value for `received_by_gex_at`.
![image](https://github.com/user-attachments/assets/2345c207-d1cf-4d4f-81f1-c50cfa79e482)
15. Let's act like we got an 824 and it had some errors. Change the status back to EDI ERROR, meaning we got the 824 back, processed it, and found an error.
16. `Important`: go to the edi_errors table, and find the entry with 824 as the edi type. Update the payment request id for that row to be the same payment request id as the 858 row, as we have been looking at one payment request.
17. Refresh your page, expand the error details, and make sure you now see 824 instead of the 858 - the 824 has a more recent created_at date, so that's the one we want to show on the UI.
![image](https://github.com/user-attachments/assets/06c4a4f5-e48b-4cab-8241-97a2c20b4446)


## Screenshots
![image](https://github.com/user-attachments/assets/a2457efe-9123-4931-86ad-912b597d5536)
